ngmix/ksigmamom.py

```diff
@@ -0,0 +1,445 @@
+            if obs.image.shape[0] > psf_obs.image.shape[0]:
+                target_dim = int(obs.image.shape[0] * self.pad_factor)
+            else:
+                target_dim = int(psf_obs.image.shape[0] * self.pad_factor)
+        else:
+            target_dim = int(obs.image.shape[0] * self.pad_factor)
+        eff_pad_factor = target_dim / obs.image.shape[0]
+
+        # pad image, psf and weight map, get FFTs, apply cen_phases
+        kim, im_row, im_col = _zero_pad_and_compute_fft(
+            obs.image, obs.jacobian.row0, obs.jacobian.col0, target_dim,
+        )
+        fft_dim = kim.shape[0]
+
+        if psf_obs is not None:
+            kpsf_im, psf_im_row, psf_im_col = _zero_pad_and_compute_fft(
+                psf_obs.image,
+                psf_obs.jacobian.row0, psf_obs.jacobian.col0,
+                target_dim,
+            )
+        else:
+            # delta function in k-space
+            kpsf_im = np.ones_like(kim, dtype=np.complex128)
+            psf_im_row = 0.0
+            psf_im_col = 0.0
+
+        # the final, deconvolved image we want is
+        #
+        #  deconv_im = kim * im_cen_phase / (kpsf_im * psf_imcen_phase)
+        #
+        # For efficiency we combine the phase comps to reduce sin and cos calls
+        # like this
+        #
+        #  deconv_im = kim / kpsf_im * (im_cen_phase / psf_im_cen_phase)
+        #
+        # The phases are complex exponentials
+        #
+        #  exp(ik*cen)
+        #
+        # So we can compute one phase as
+        #
+        #  im_cen_phase / psf_im_cen_phase = exp(ik * (im_cen - psf_cen))
+        #
+        # and then multiply it into the image.
+        #
+        # This operation and the deconvolutiomn will be done
+        # later in _measure_moments_fft
+
+        # now build the kernels
+        kres = _ksigma_kernels(
+            target_dim,
+            self.fwhm,
+            obs.jacobian.dvdrow, obs.jacobian.dvdcol,
+            obs.jacobian.dudrow, obs.jacobian.dudcol,
+        )
+
+        # compute the total variance from weight map
+        msk = obs.weight > 0
+        tot_var = np.sum(1.0 / obs.weight[msk])
+
+        # run the actual measurements and return
+        res = _measure_moments_fft(
+            kim, kpsf_im, tot_var, eff_pad_factor, kres,
+            im_row - psf_im_row, im_col - psf_im_col,
+        )
+        if res['flags'] != 0:
+            logger.debug("ksigma pre-psf moments failed: %s" % res['flagstr'])
+
+        if return_kernels:
+            # put the kernels back into their unpacked state
+            new_kres = {}
+            for k in kres:
+                if k == "msk":
+                    continue
+                new_kres[k] = np.zeros((fft_dim, fft_dim), dtype=np.complex128)
+                new_kres[k][kres["msk"]] = kres[k]
+            res["kernels"] = new_kres
+
+        return res
+
+
+def _measure_moments_fft(kim, kpsf_im, tot_var, eff_pad_factor, kernels, drow, dcol):
```

Comment:
I think this function could be refactored, e.g. pack the results into a dict in a separate function.  But not big deal